### PR TITLE
Support multiple UPnP devices

### LINKED ISSUE #787: Port mapping fails if the first UPnP device found is on a different subnet
### What is wrong?

The server grabs the first device it can find and runs with it. So if the UPnP device on the same `/24` address space is the second one listed, NAT port mapping fails.

### How can it be fixed

If the first device mapping fails, keep iterating through the devices until one succeeds (or you run out of devices).

I have a very dirty PR on my machine. I haven't pushed it yet because it's not critical path to get the alpha out the door.

### PR DESCRIPTION
**Pull request by carver**
### What was wrong?

If the first UPnP device you found was not working (say, because it was on a different subnet), then we just gave up.

Fixes #787 

### How was it fixed?

- If setting up NAT on the first device fails, attempt to set up on the others
- Extracted UPnP logic out of p2p.server

Bonus fix:
- bugfix when canceling an already canceled service (now it will only warn instead of raise an exception)

#### Cute Animal Picture

put a cute animal picture link inside the parentheses

### CODE REVIEW

File: p2p/nat.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+# UPnP discovery can take a long time, so use a loooong timeout here.
+UPNP_DISCOVER_TIMEOUT_SECONDS = 30
+
+
+class PortMapping(NamedTuple):
+    internal: str  # of the form "192.2.3.4:56"
+    external: str  # of the form "192.2.3.4:56"
+
+
+class NoInternalAddressMatchesDevice(Exception):
```
Comment (pipermerriam) [inline p2p/nat.py]: Is there a reason this isn't in `p2p.exceptions`?

File: p2p/server.py [hunk 2]
```diff
@@ -246,11 +142,11 @@ def _make_peer_pool(self, discovery: DiscoveryProtocol) -> PeerPool:
...
     async def _run(self) -> None:
         self.logger.info("Running server...")
-        upnp_dev = await self._discover_upnp_device()
-        external_ip = '0.0.0.0'
-        if upnp_dev is not None:
-            external_ip = upnp_dev.WANIPConn1.GetExternalIPAddress()['NewExternalIPAddress']
-            await self._add_nat_portmap(upnp_dev)
+        mapped_external_ip = await self.upnp_service.add_nat_portmap()
+        if mapped_external_ip is None:
+            external_ip = '0.0.0.0'
```
Comment (pipermerriam) [inline p2p/server.py]: Wondering if we can move away from this and instead do what geth does and just not show an IP address.

File: p2p/nat.py (hunk 1 above)
Comment (carver) [reply p2p/nat.py]: Everything in `p2p.exceptions` seems to be related to peers, and this doesn't really have anything to do with it. Maybe `nat` should be moved to `trinity`.

Comment (pipermerriam) [reply p2p/nat.py]: I think it is safe to assume anyone who wants to setup an app which uses the DevP2P protocol will want this functionality so `p2p` seems more appropriate.

**Pull request merged**